### PR TITLE
build: Update to MN 4

### PR DESCRIPTION
**Pull request by sdelamo**

### CODE REVIEW

**Comment by sdelamo** [comment]
I merged into master upgrades to Spring 3 and Spring Boot 6. I anticipate we will need to do several changes in this PR. 

Can you work on this @timyates ?

File: spring-context/src/main/java/io/micronaut/spring/context/MicronautApplicationContext.java [hunk 1]
```diff
@@ -380,6 +386,11 @@ public void addApplicationListener(ApplicationListener<?> listener) {
         beanFactory.getBeanContext().registerSingleton(listener);
     }
 
+    @Override
+    public void removeApplicationListener(ApplicationListener<?> listener) {
+        beanFactory.getBeanContext().destroyBean(listener);
```
Comment (sdelamo) [inline spring-context/src/main/java/io/micronaut/spring/context/MicronautApplicationContext.java:391]: Is this the correct way to remove the listener singleton?

File: spring-context/src/main/java/io/micronaut/spring/context/factory/MicronautBeanFactory.java [hunk 2]
```diff
@@ -460,7 +460,9 @@ public Class<?> getType(@NonNull String beanName) throws NoSuchBeanDefinitionExc
                     if (beanClass != null && FactoryBean.class.isAssignableFrom(beanClass)) {
                         if (!BeanFactoryUtils.isFactoryDereference(beanName)) {
                             // If it's a FactoryBean, we want to look at what it creates, not at the factory class.
-                            return super.getTypeForFactoryBean(beanName, mbd);
+                            boolean allowInit = true; // if allowInit is true a full creation of the FactoryBean is used as fallback (through delegation to the superclass's implementation).
+                            ResolvableType resolvableType = super.getTypeForFactoryBean(beanName, mbd, allowInit);
+                            return resolvableType.getRawClass();
```
Comment (sdelamo) [inline spring-context/src/main/java/io/micronaut/spring/context/factory/MicronautBeanFactory.java:465]: not sure if this is correct. There is no method which takes only two parameters `beanName`,`mbd` any more

File: spring-web/src/main/java/io/micronaut/spring/web/reactive/MicronautServerHttpRequest.java [hunk 3]
```diff
@@ -120,11 +126,12 @@ public reactor.core.publisher.Flux<DataBuffer> getBody() {
...
             final NettyDataBufferFactory nettyDataBufferFactory = new NettyDataBufferFactory(channel.alloc());
 
-            final Optional<HttpContentProcessor<ByteBufHolder>> httpContentProcessor = channelResolver.resolveContentProcessor(request);
+            final Optional<HttpContentProcessor> httpContentProcessor = channelResolver.resolveContentProcessor(request);
 
             if (httpContentProcessor.isPresent()) {
 
-                final HttpContentProcessor<ByteBufHolder> processor = httpContentProcessor.get();
+                final HttpContentProcessor processor = httpContentProcessor.get();
+                /*
```
Comment (sdelamo) [inline spring-web/src/main/java/io/micronaut/spring/web/reactive/MicronautServerHttpRequest.java:134]: not sure how this should be implemented

File: spring/src/main/java/io/micronaut/spring/core/annotation/MergedAnnotationValue.java [hunk 4]
```diff
@@ -59,7 +59,9 @@ public MergedAnnotationValue(AnnotationMetadata annotationMetadata, AnnotationVa
     @Override
     public Class<A> getType() {
         return (Class<A>) annotationMetadata.getAnnotationType(value.getAnnotationName())
-                    .orElseGet(() -> ClassUtils.forName(value.getAnnotationName(), getClass().getClassLoader()).orElseThrow(() -> new IllegalStateException("Annotation class not present")));
+                    .orElseGet(() -> (Class<A>) ClassUtils.forName(value.getAnnotationName(), getClass().getClassLoader())
```
Comment (sdelamo) [inline spring/src/main/java/io/micronaut/spring/core/annotation/MergedAnnotationValue.java:62]: I had to add the cast

File: spring/src/main/java/io/micronaut/spring/core/type/BeanDefinitionSpringMetadata.java [hunk 5]
```diff
@@ -47,6 +47,11 @@ public Set<MethodMetadata> getAnnotatedMethods(String annotationName) {
         return Collections.emptySet();
     }
 
+    @Override
+    public Set<MethodMetadata> getDeclaredMethods() {
+        return Collections.emptySet();
```
Comment (sdelamo) [inline spring/src/main/java/io/micronaut/spring/core/type/BeanDefinitionSpringMetadata.java:52]: not sure how this method should be implemented?

File: spring/src/main/java/io/micronaut/spring/core/type/ClassElementSpringMetadata.java [hunk 6]
```diff
@@ -51,6 +52,11 @@ public Set<MethodMetadata> getAnnotatedMethods(String annotationName) {
             .collect(Collectors.toSet());
     }
 
+    @Override
+    public Set<MethodMetadata> getDeclaredMethods() {
+        return Collections.emptySet();
```
Comment (sdelamo) [inline spring/src/main/java/io/micronaut/spring/core/type/ClassElementSpringMetadata.java:57]: not sure how this method should be implemented

File: spring-context/src/main/java/io/micronaut/spring/context/factory/MicronautBeanFactory.java (hunk 2 above)
Comment (graemerocher) [reply spring-context/src/main/java/io/micronaut/spring/context/factory/MicronautBeanFactory.java:465]: probably fine

File: spring-context/src/main/java/io/micronaut/spring/context/MicronautApplicationContext.java (hunk 1 above)
Comment (graemerocher) [reply spring-context/src/main/java/io/micronaut/spring/context/MicronautApplicationContext.java:391]: looks ok

**Comment by graemerocher** [comment]
not sure why checkstyle is failing

**Pull request merged**